### ISSUE TITLE
pytree transformation

### ISSUE THREAD

**Issue by shoyer** [google/jax#1012]
I'd like to write code like `np.sum((x - y) ** 2)` that automatically works on arbitrarily nested Python objects.

I think this could be done cleanly with a "pytree transformation" that acts on functions, somewhat similar to the existing `vmap`, e.g.,
```
@jax.autotree
def norm(x, y):
    return np.sum((x - y) ** 2)
```

Under the hood, we would need to define what "autotree" versions of each primitive function look like on a list of heterogeneously shaped arrays, e.g.,
- Vectorized functions like `+` operate over each array independently.
- In mixed tree/array operations, the array should get "broadcast" by repeating it when the structure of the tree, e.g., `{'a': x, 'b': y'} + z` -> `{'a': x + z, 'b': y + z}`.
- Reductions like `sum` sum over all the arrays to produce a scalar, unless an axis is specified.

The upside is that this could make it trivial to write functional transformations like `jax.experimental.optimizers` in a fully general way. I can think of lots of nice applications for this, e.g., for methods that solve equations.

**Comment by mattjj** [comment]
This doesn't solve the full ("applicative"?) case you want, but I thought I'd mention it:

```python
from functools import partial
import jax.numpy as np
from jax.tree_util import tree_multimap

@partial(partial, tree_multimap)
def norm(x, y):
  return np.sum((x - y) ** 2)  # actually a _squared_ norm ;)

x = y = (np.ones(3), [np.ones(3, 4), np.ones(5)])
print(norm(x, y))  # (0., [0., 0.])
```

This is a way to implement a pure map over trees (analogous to vmap), but doesn't support reductions (like pmap does with special collectives).

Since jaxprs don't see pytrees anyway, and see all product-type-like structures as flat multiple-inputs / multiple-outputs, maybe we can simplify the problem to just be about tuples, or even just pairs. That is, we want to transform this function (or rather its jaxpr)

```python
def norm(x, y):
  return np.sum((x - y) ** 2)
```

```
{ lambda  ;  ; a b.
  let c = sub a b
      d = pow c 2.0
      e = reduce_sum[ axes=(0,)
                      input_shape=(2,) ] d
  in [e] }
```

into this one (or rather its jaxpr)

```python
def norm(x1, x2, y1, y2):
  norm1 = np.sum((x1 - y1) ** 2)
  norm2 = np.sum((x2 - y2) ** 2)
  return norm1 + norm2
```

```
{ lambda  ;  ; a b c d.
  let e = sub a c
      f = pow e 2.0
      g = reduce_sum[ axes=(0,)
                      input_shape=(2,) ] f
      h = sub b d
      i = pow h 2.0
      j = reduce_sum[ axes=(0,)
                      input_shape=(2,) ] i
      k = add g j
  in [k] }
```

Seems like something that can be taken up more easily now that we have a [colab notebook on jaxpr interpreters](https://github.com/google/jax/pull/1582)!

**Comment by shoyer** [comment]
> Since jaxprs don't see pytrees anyway, and see all product-type-like structures as flat multiple-inputs / multiple-outputs, maybe we can simplify the problem to just be about tuples, or even just pairs. That is, we want to transform this function (or rather its jaxpr)

I think this could probably be done just at the primitive level, which is how JAX's existing transformations work. Like JAXprs, primitives take and return flat lists of arguments.

That said, I don't know if there's any way we could automatically handle the interesting cases like `np.sum`. The JAXpr for `sum` shows that the primitive includes axes, even if you don't specify them directly:
```
>>> jax.make_jaxpr(np.sum)(np.ones((2, 3)))
{ lambda  ;  ; a.
  let b = reduce_sum[ axes=(0, 1)
                      input_shape=(2, 3) ] a
  in [b] }
```

**Comment by mattjj** [comment]
Why is the original jaxpr containing axes a problem?

**Comment by shoyer** [comment]
> Why is the original jaxpr containing axes a problem?

Only because I wanted to distinguish between the behavior of `sum(x)` and `sum(x, axis=tuple(range(x.ndim)))`

Instead, I think this sort of thing should be explicitly marked in some way, e.g., 
```
@jax.autotree
def norm(x, y):
    return np.sum(flatten((x - y) ** 2))
```
where `flatten()` is a no-op except under an "auto-tree" transformation, in which case it flattens and concatenates all the inputs.

**Comment by mattjj** [comment]
One clear way to think about this is to ask: how would you write this in [Dex](https://github.com/google-research/dex-lang)? @dougalm 

Since we're expressing more than just pure maps over trees (i.e. also reductions), we might want to take a page or two from `pmap`'s design.

**Comment by shoyer** [comment]
For a relatively simple use-case, take a look at https://github.com/google/jax/pull/2600 which generalizes `cg` to work on Pytrees.

Before:
```python
  def body_fun(value):
    x, r, gamma, p, k = value
    Ap = A(p)
    alpha = gamma / _vdot(p, Ap)
    x_ = x + alpha * p
    r_ = r - alpha * Ap
    z_ = M(r_)
    gamma_ = _vdot(r_, z_)
    beta_ = gamma_ / gamma
    p_ = z_ + beta_ * p
    return x_, r_, gamma_, p_, k + 1
```
After (much less readable!):
```python
  def body_fun(value):
    x, r, gamma, p, k = value
    Ap = A(p)
    alpha = gamma / _vdot(p, Ap)
    x_ = _add(x, _mul(alpha, p))
    r_ = _sub(r, _mul(alpha, Ap))
    z_ = M(r_)
    gamma_ = _vdot(r_, z_)
    beta_ = gamma_ / gamma
    p_ = _add(z_, _mul(beta_, p))
    return x_, r_, gamma_, p_, k + 1
```

**Comment by shoyer** [comment]
Take a look at #3263 for preliminary implementation. The key insight (not sure how I missed this before) is that we can write a Tracer that represents an arbitrary tree of arrays as if it were a flattened/concatenated vector.

**Comment by gehring** [comment]
I just want to show my support for this feature. Recently, I've been implementing things like `gmres` and `cg` and supporting pytrees is always a hassle. I usually start by writing up a non-pytree version and it always breaks my heart to see how less readable the code is once I've added support for pytrees.

I wouldn't consider this feature necessary but it's definitely a common source of friction and an equally good source of bugs. If jax could support something like this without making everything significantly harder to maintain, it would be a big win in my book!

**Comment by NeilGirdhar** [comment]
I just came to post this feature request.

Now that L-BFGS has been checked in, I wanted to use it with pytrees, and I first thought of contributed that change, but then I realized that this is a more general problem.  Often, it's easier to code things to work for arrays, and it seems like by the separation of concerns, you'd want the extension to PyTrees to be done somewhere else if possible.

Would it make more sense for the definition to support choosing which parameters are identically-structured trees?  For example, something like this:
```python
from jax.scipy.optimize import minimize
from jax import autotree


minimize = autotree(in_trees=(None, 0, None),
                out_trees=OptimizeResults(0, None, None, None, None, None, None, None, None))(
                    minimize)

minimize(fun, x0, args)
```
This is similar to `vmap` in that the same number represents identically-structured trees.  Multiple numbers don't need to be supported since you could apply autotree multiple times (once for each identically-structured tree).

The original example at the top of this issue would be:
```python
@jax.autotree(in_trees=(0, 0), out_trees=0
def norm(x, y):
    return np.sum((x - y) ** 2)
```

What do you think?

**Comment by shoyer** [comment]
I need to figure out how to get #3263 moving towards something we can merge. The main unresolved challenge to make this useful is figuring out how to get JAX transformations and control flow inside tree vectorized functions to work. There is a lot of reorganizing data for new calling conventions that gets rather tedious, but maybe we can just power through it or figure out a better way, e.g., perhaps with partial tree flattening using the new(ish) `is_leaf` parameter.

With regard to `in_trees`/`out_trees`, that seems like a very sensible suggestion (similar to `in_axes`/`out_axes` for `vmap`), although you can get a similar effect to `in_trees` just by defining an inner function to apply `tree_vectorize` to using a closure. I think this is somewhat more of a superficial issue, though -- the key challenge to tackle first is figuring out how to get the fundamental JAX transformations inside `tree_vectorize` working.

CC @apaszke

**Comment by shoyer** [comment]
Update: I have a new approach for implementing this, which I think is much more promising than #3263:
- #8504

Feedback would be very welcome!

**Comment by shoyer** [comment]
This has now been implemented in a separate library, `tree-math`: https://github.com/google/tree-math/
